apps/communications/contacts/views/details/js/details_controller.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+/* global utils */
+/* global MatchService */
+/* global ContactsService */
+/* global LazyLoader */
+/* global ContactToVcardBlob */
+/* global VcardFilename */
+/* global MozActivity */
+
+/* exported Details */
+
+(function(exports) {
+
+  var _activity = null;
+
+  function setActivity(activity) {
+    _activity = activity;
+  }
+
+  function findDuplicates(evt) {
+    if (!evt.detail || !evt.detail.contactId) {
+      console.error('Missing parameters in CustomEvent');
+      return;
+    }
+
+    var contactId = evt.detail.contactId;
+    MatchService.match(contactId);
+  }
+
+  function toggleFavorite(evt){
+    if (!evt.detail || typeof evt.detail.contact === 'undefined' ||
+      typeof evt.detail.isFavorite === 'undefined') {
+      console.error('Missing parameters in CustomEvent');
+      return;
+    }
+
+    var contact = evt.detail.contact;
+    var isFavorite = evt.detail.isFavorite;
+    var favorite = !isFavorite;
+    if (favorite) {
+      contact.category = contact.category || [];
+      contact.category.push('favorite');
+    } else {
+      if (!contact.category) {
+        return;
+      }
+      var pos = contact.category.indexOf('favorite');
+      if (pos > -1) {
+        contact.category.splice(pos, 1);
+      }
+    }
+
+    function onError(e) {
+      console.error('Error saving favorite');
+      // We must restore contact in order to update the UX accordingly
+      if (favorite) {
+        var pos = contact.category.indexOf('favorite');
+        if (pos > -1) {
+          contact.category.splice(pos, 1);
+        }
+      } else {
+        contact.category = contact.category || [];
+        contact.category.push('favorite');
+      }
+
+      dispatchEvent('toggleFavoriteDone', {contact: contact});
+    }
+
+    ContactsService.save(
+      utils.misc.toMozContact(contact),
+      function(e) {
+        if (typeof e !== 'undefined') {
+          onError(e);
+          return;
+        }
+
+        ContactsService.get(contact.id, function(savedContact) {
+          dispatchEvent('toggleFavoriteDone', {contact: savedContact});
+        }, onError);
+      }
+    );
+  }
+
+  function init(activity) {
+    if (typeof activity !== 'undefined') {
+      setActivity(activity);
+    }
+
+    window.addEventListener('backAction', handleBackAction);
+    window.addEventListener('editAction', handleEditAction);
+    window.addEventListener('toggleFavoriteAction', toggleFavorite);
+    window.addEventListener('shareAction', shareContact);
+    window.addEventListener('findDuplicatesAction', findDuplicates);
+
+    navigator.mozContacts.oncontactchange = evt => {
```

Comment:
I would remove this by now, due to it's not going to be used out of the activity. We will add this later on, when adding this view to the main App. Please add a `TODO` with the explanation, just to have a clear picture of what we are implementing here and the pending work to be done.